src/libraries/Common/src/Interop/OSX/Interop.libobjc.cs

```diff
@@ -28,16 +28,16 @@ private struct NSOperatingSystemVersion
         private static extern IntPtr objc_getClass(string className);
         [DllImport(Libraries.libobjc)]
         private static extern IntPtr sel_getUid(string selector);
-        [DllImport(Libraries.libobjc)]
-        private static extern IntPtr objc_msgSend(IntPtr basePtr, IntPtr selector);
+        [DllImport(Libraries.libobjc, EntryPoint = "objc_msgSend")]
+        private static extern IntPtr get_ProcessInfo(IntPtr basePtr, IntPtr selector);
```

Comment:
This does not follow https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/interop-guidelines.md#pinvoke-definitions

Reply:
But it is impossible to strictly follow that guidance for this API given how it works. All Objective-C uses of `objc_msgSend` need to be special cased in practice since the name is simply the generic symbol for all message functions. Applying this logic all places in the runtime that need `objc_msgSend` would have the same name but different signatures. Having different signatures for the same function seems far more confusing than applying a clear name and marking the entry point as appropriate.

Reply:
The method that it is going to call is passed in as the second `objc_msgSend` argument. I do not think that it is confusing.

libraries interop has other similar methods that can do many things, one of the argument says which one of the many things the method should do, and it changes meaning of the remaining arguments. We have been following the naming guidance for these methods so far.

Reply:
Then how do we square this usage with usage in the same file further below:

https://github.com/dotnet/runtime/blob/42bc286e441f250843cd9edb67e25129436f13a5/src/libraries/Common/src/Interop/OSX/Interop.libobjc.cs#L66-L67

Or with the [Xamarin guidance](https://docs.microsoft.com/xamarin/ios/internals/objective-c-selectors) on how to use `objc_msgSend`.

The workaround later in the file coupled with the Xamarin guidance seem to indicate a consensus that specificity in this case should be the priority. I concur that if this method's signature was consistent there would be little argument. However, this is the only case I can find where the [declared signature](https://developer.apple.com/documentation/objectivec/1456712-objc_msgsend) in doc - was made `void(void)` to avoid this confusion - is explicitly ignored in a P/Invoke signature. That would mean any existing P/Invoke guidance in the runtime shouldn't apply and rather deferring to community accepted policy would be appropriate.

If the above doesn't resonate, then we can drop this. My personal preference was to do all of this in Objective-C but there are sound arguments against that.

Reply:
Libraries interop naming convention suggests to append `_Something` to disambiguate as necessary. For example, we have `IntPtr CreateFile_IntPtr` to disambiguate it with `SafeHandle CreateFile`. It is similar to Xamarin naming guidance, except that the disambiguation with the Xamarin guidance is non-optional and uses all arguments.

Using Xamarin naming guidance (include types in the DllImport name) or some less verbose version of it (e.g. only include return type) sounds reasonable to me.

Reply:
I'm not familiar with ObjC, but what's stopping us from having a single `objc_msgSend` P/Invoke and then a wrapper function named `get_ProcessInfo` that calls it, for example?

Reply:
`objc_msgSend` does not have a fixed signature. It is similar to `printf` where the first argument describes what the rest signature actually is. The differences between `printf` and `objc_msgSend` are:
- `objc_msgSend` uses regular calling convention, prinftf uses vararg calling convention
- Return type can vary for `objc_msgSend`, it is always `void` for printf

Reply:
Is this maybe a scenario that would be a (low priority) driver for supporting generic P/Invokes

Reply:
>Is this maybe a scenario that would be a (low priority) driver for supporting generic P/Invokes

I don't think so given the entire signature is at odds. Not just types but number of arguments. This also is a niche pattern (i.e. interop with message passing languages Objective-C, SmallTalk, etc) so I'm not sure it would be an argument to drive that support.